### PR TITLE
Add ability to speak next and previous lines/paragraphs when navigating with braille

### LINKED ISSUE #17053: braille: optionally speak line when moving from line to line
### Is your feature request related to a problem? Please describe.
Likely many braille users in many cases use both braille and speech simultaneously. Thus it would be useful when `nextLine` and `previousLine` functions are executed that optionally also new line would be spoken.
Firstly I thought this would be useful when braille is tethered to focus but now I think it should be available when braille is tethered to review as well.

### PR DESCRIPTION
**Pull request by nvdaes**
### Link to issue number:
Fixes #17053 
### Summary of the issue:
Sometimes, it may be desirable tohear lines when navigating with braille.
### Description of user facing changes
- A new command, not assigned to gestures, is available in the braille category to toggle speaking of lines/paragraphs when navigating with braille, to enable or disable this feature. A new checkbox for this has been added in the braille settings panel as well.

### Description of development approach
- A new key for this feature has been added to configSpec.py. The value is a boolean (False by default).
- A new script has been added to globalCommands.py.
- A new function has been added to braille.py, to be used by next and previousLine functions, also modified so that info is expanded to the corresponding read unit, i.e., line or paragraph.
- The braille settings panel has a new checkbox to toggle this feature.
### Testing strategy:
Tested manually in Notepad and Window console, with braille following caret, review cursor and automatic modes.
Also, checked that indentation is reported when reading by lines, not by paragraphs, as expected when unit is specified for speakTextInfo function.
### Known issues with pull request:
None.
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

@coderabbitai summary

### CODE REVIEW

**Comment by nvdaes** [comment]
@burmancomp , feel free to test this again. I've added `cancelSpeech()` before speaking line or paragraph, and a checkbox in the braille settings panel.

**Comment by nvdaes** [comment]
I updated the changelog and user guide.
I'll test once more and then I'll request a review from NV Access.

**Comment by nvdaes** [comment]
I feel that changelog.md and userGuide.md are missing.

Thanks @burmancomp . I've pushed changes correctly now and this should be fixed and ready for review.

**Comment by nvdaes** [comment]
@burmancomp , happy to see that this is useful for you.

File: user_docs/en/userGuide.md [hunk 1]
```diff
@@ -2297,6 +2297,12 @@ If this is enabled, NVDA will automatically speak the character at the cursor wh
 
 To toggle this option from anywhere, please assign a custom gesture using the [Input Gestures dialog](#InputGestures).
 
+##### Speak character when navigating by line or paragraph {#BrailleSpeakOnNavigating}
+
+If this is enabled, NVDA will automatically speak the current line or paragraph when navigating to it with braille navigation keys.
```
Comment (CyrilleB79) [inline user_docs/en/userGuide.md:2302]: Slight rewording that seems better to me, although that cans probably still be improved.

```suggestion
If this is enabled, NVDA will speak the reached line or paragraph when moving the braille display to the previous or next line.
```

File: source/gui/settingsDialogs.py [hunk 2]
```diff
@@ -4613,6 +4613,14 @@ def makeSettings(self, settingsSizer):
 		self.bindHelpEvent("BrailleSpeakOnRouting", self.speakOnRoutingCheckBox)
 		self.speakOnRoutingCheckBox.Value = config.conf["braille"]["speakOnRouting"]
 
+		# Translators: The label for a setting in braille settings to speak the current line or paragraph when navigating by them with braille.
+		speakOnNavigatingText = _("Speak &line or paragraph when navigating by them with braille")
```
Comment (CyrilleB79) [inline source/gui/settingsDialogs.py:4617]: IMO, the option could be reworded.
* No need to tell "word or paragraph": the feedback depends on the corresponding option
* No need to mention braille since we are already in braille panel IMO

```suggestion
		speakOnNavigatingText = _("Provide speech feedback when moving by &line")
```

**Review by CyrilleB79 — COMMENTED** [review]
Some rewording suggestions; this may be discussed.

**Comment by nvdaes** [comment]
Thanks @CyrilleB79 . I've applied your suggestions.

File: user_docs/en/userGuide.md [hunk 3]
```diff
@@ -2297,6 +2297,12 @@ If this is enabled, NVDA will automatically speak the character at the cursor wh
 
 To toggle this option from anywhere, please assign a custom gesture using the [Input Gestures dialog](#InputGestures).
 
+##### Speak when navigating by line or paragraph {#BrailleSpeakOnNavigating}
+
+If this is enabled, NVDA will speak the reached line or paragraph when moving the braille display to the previous or next line.
```
Comment (XLTechie) [inline user_docs/en/userGuide.md:2302]: Using "reached" with this sentence structure, sounds a bit irregular in English. Maybe try:

```suggestion
Enabling this option will cause NVDA to speak lines or paragraphs reached using the braille display's navigation controls.
```

At least, given my understanding of the option as essentially "speech follows braille".

**Comment by nvdaes** [comment]
Thanks @XLTechie . I've applied your suggestion.

File: source/braille.py [hunk 4]
```diff
@@ -1659,6 +1659,8 @@ def nextLine(self):
 				return
 		dest.collapse()
 		self._setCursor(dest)
+		dest.expand(self._getReadingUnit())
```
Comment (LeonarddeR) [inline source/braille.py:1662]: I'm afraid this can become problematic, since you're mutating the TextInfo that was said as cursor before. Instead, I think you should copy the textInfo first.
```suggestion
		copy = dest.copy()
		copy.expand(self._getReadingUnit())
```

File: source/braille.py [hunk 5]
```diff
@@ -1677,11 +1679,13 @@ def previousLine(self, start=False):
 					pass
 				else:
 					dest = dest.obj.makeTextInfo(textInfos.POSITION_LAST)
-					dest.expand(unit)
-			else:  # no page turn support
+			else:
+				# no page turn supportcode ..\nvda
```
Comment (LeonarddeR) [inline source/braille.py:1683]: It looks like some mess ended up in here
```suggestion
				# no page turn support
```

File: source/braille.py [hunk 6]
```diff
@@ -1677,11 +1679,13 @@ def previousLine(self, start=False):
 					pass
 				else:
 					dest = dest.obj.makeTextInfo(textInfos.POSITION_LAST)
-					dest.expand(unit)
```
Comment (LeonarddeR) [inline source/braille.py:1680]: Why was this removed?

File: source/braille.py [hunk 7]
```diff
@@ -1677,11 +1679,13 @@ def previousLine(self, start=False):
...
-			else:  # no page turn support
+			else:
+				# no page turn supportcode ..\nvda
 				return
 		dest.collapse()
 		self._setCursor(dest)
+		dest.expand(self._getReadingUnit())
```
Comment (LeonarddeR) [inline source/braille.py:1687]: Please copy first.

File: source/braille.py [hunk 8]
```diff
@@ -3841,3 +3845,18 @@ def _speakOnRouting(info: textInfos.TextInfo):
 
 	info.expand(textInfos.UNIT_CHARACTER)
 	spellTextInfo(info)
+
+
+def _speakOnNavigatingByUnit(info: textInfos.TextInfo):
```
Comment (LeonarddeR) [inline source/braille.py:3850]: If this function is supposed to speak the reading unit, I think I'd do the expansion in here rather than in the methods that call it. May be even do the copying in here to be on the safe side of things.

File: source/braille.py [hunk 9]
```diff
@@ -3841,3 +3845,18 @@ def _speakOnRouting(info: textInfos.TextInfo):
...
+	"""Speaks the reading unit after navigating by it with braille.
+
+	:param info: The reading unit TextInfo at the cursor position after navigating.
+	"""
+	if not config.conf["braille"]["speakOnNavigatingByUnit"]:
+		return
+	# Import late to avoid circular import.
+	from speech.speech import cancelSpeech, speakTextInfo
+
+	readingUnit = handler.buffer.regions[-1]._getReadingUnit()
```
Comment (LeonarddeR) [inline source/braille.py:3860]: Rather than getting the actual unit this way, I'd add it as a parameter to `_speakOnNavigatingByUnit` instead.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Great to see this being implemented!

File: source/braille.py (hunk 6 above)
Comment (nvdaes) [reply source/braille.py:1680]: Accidentally. Thanks for pointing it out.

**Comment by nvdaes** [comment]
Thanks @LeonarddeR for your review. I'll mark this as a draft and I'll work on this later.

**Comment by nvdaes** [comment]
@LeonarddeR , I think that I've addressed all your comments. I don't have if you can review this again, so I'll mark this as ready for review in case NV Access wants to do it.

**Comment by nvdaes** [comment]
@burmancomp , I'll look at this tomorrow.

File: source/braille.py [hunk 10]
```diff
@@ -3841,3 +3844,20 @@ def _speakOnRouting(info: textInfos.TextInfo):
...
 	info.expand(textInfos.UNIT_CHARACTER)
 	spellTextInfo(info)
+
+
+def _speakOnNavigatingByUnit(info: textInfos.TextInfo, readingUnit: str) -> None:
+	"""Speaks the reading unit after navigating by it with braille.
+
+	:param info: The TextInfo at the cursor position after navigating.
+	:param readingUnit: The reading unit to expand TextInfo.
+	"""
```
Comment (SaschaCowley) [inline source/braille.py:3854]: Can the docstring be expanded to note that this won't do anything if the user has disabled this feature?
```suggestion
	"""Speaks the reading unit after navigating by it with braille.
	
	This only has an effect if the user has enabled "Speak when navigating by line or paragraph" in braille settings.

	:param info: The TextInfo at the cursor position after navigating.
	:param readingUnit: The reading unit to expand TextInfo.
	"""
```

File: source/braille.py [hunk 11]
```diff
@@ -3841,3 +3844,20 @@ def _speakOnRouting(info: textInfos.TextInfo):
...
+	if not config.conf["braille"]["speakOnNavigatingByUnit"]:
+		return
+	# Import late to avoid circular import.
+	from speech.speech import cancelSpeech, speakTextInfo
+
+	cancelSpeech()
+	copy = info.copy()
+	copy.expand(readingUnit)
+	speakTextInfo(copy, unit=readingUnit, reason=controlTypes.OutputReason.CARET)
```
Comment (SaschaCowley) [inline source/braille.py:3863]: Is there a reason to `cancelSpeech` before copying and expanding the `TextInfo`, rather than immediately before `speakTextInfo`?
```suggestion
	copy = info.copy()
	copy.expand(readingUnit)
	cancelSpeech()
	speakTextInfo(copy, unit=readingUnit, reason=controlTypes.OutputReason.CARET)
```

**Review by SaschaCowley — CHANGES_REQUESTED** [review]
Thanks @nvdaes, this is looking really good.

**Review by SaschaCowley — APPROVED** [review]
Thanks for this good work, @nvdaes

**Comment by nvdaes** [comment]
Thanks for your work and your review, @SaschaCowley .

File: user_docs/en/userGuide.md [hunk 12]
```diff
@@ -2297,6 +2297,12 @@ If this is enabled, NVDA will automatically speak the character at the cursor wh
 
 To toggle this option from anywhere, please assign a custom gesture using the [Input Gestures dialog](#InputGestures).
 
+##### Speak when navigating by line or paragraph {#BrailleSpeakOnNavigating}
+
+Enabling this option will cause NVDA to speak lines or paragraphs reached using the braille display's navigation controls.
+
+To toggle this option from anywhere, please assign a custom gesture using the [Input Gestures dialog](#InputGestures).
```
Comment (Qchristensen) [inline user_docs/en/userGuide.md:2304]: While this is not dissimilar to how other commands are written, I know users sometimes have difficulty finding the exact gestures to assign in the input gesture dialog, so I'd like to propose:

```suggestion
To toggle this option from anywhere, please assign a custom gesture to "speakOnNavigatingByUnit" in the "Braille" section of the [Input Gestures dialog](#InputGestures).
```

**Review by Qchristensen — CHANGES_REQUESTED** [review]
Looks good - one suggestion proposed as a way to improve the user guide moving forward

**Comment by nvdaes** [comment]
Thanks @Qchristensen . I've applied your suggestion.

**Review by Qchristensen — APPROVED** [review]
